conftest.py

```diff
@@ -50,11 +96,36 @@ def pytest_collection_modifyitems(config, items):
             )
             item.add_marker(marker)
 
+    run_network_tests = environ.get('SKLEARN_SKIP_NETWORK_TESTS', '1') == '0'
+    skip_network = pytest.mark.skip(
+        reason="test requires internet connectivity")
+
+    # download datasets during collection to avoid thread unsafe behavior
+    # when running pytest in parallel with pytest-xdist
+    dataset_features_set = set(dataset_fetchers)
+    datasets_to_download = set()
+
+    for item in items:
+        item_keywords = set(item.keywords)
+        dataset_to_fetch = item_keywords & dataset_features_set
+        if not dataset_to_fetch:
+            continue
+
+        if run_network_tests:
+            datasets_to_download |= dataset_to_fetch
+        else:
+            # network tests are skipped
+            item.add_marker(skip_network)
+
+    # download datasets that are needed to avoid thread unsafe behavior
+    # by pytest-xdist
+    if run_network_tests:
+        for name in datasets_to_download:
+            dataset_fetchers[name]()
+
     # Skip tests which require internet if the flag is provided
     if (config.getoption("--skip-network")
             or int(os.environ.get("SKLEARN_SKIP_NETWORK_TESTS", "0"))):
```

Comment:
The `run_network_tests` variable above is not strictly equivalent to this conditional expression. Shouldn't the 2 be the same, in which case we could just do `if not run_network_tests` instead?